activesupport/lib/active_support/message_verifier.rb

```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    # Generates a signed message for the given +value+ and +options+ hash.
     #
     # The message is signed with the +MessageVerifier+'s secret. Without knowing
-    # the secret, the original value cannot be extracted from the message.
+    # the secret, the original value cannot be extracted from the message. A
+    # message is signed in the form of JSON Web Token(JWT). The final message is
+    # concatenation of base64-encoded JWS header, base64-encoded JWT claims set
+    # and  base64-encoded signature seperated by a period('.'). This format is
+    # known as JSON Web Signature(JWS) structure:
+    # <base64-encoded header>.<base64-encoded claims>.<base64-encoded signature>
+    #
+    # JSON Web Token (JWT) - A string representing a set of claims as a JSON
+    # object that is encoded in a JWS or JWE, enabling the claims to be
+    # signed and/or encrypted.
+    #
+    # JWT Header - A JSON object that describes the cryptographic operations
+    # applied to the JWT. When the JWT is signed, the JWT Header is a
+    # JWS Header. A JWS header signed using 'SHA1' algorithm:
+    #
+    #   { 'typ' => 'JWT', 'alg' => 'SHA1' }
+    #
+    # JWT claims set - A JSON object that contains the claims conveyed by
+    # the JWT. For our purpose, we limit our claims set to consist of
+    # +pld+(payload), +exp+(expiry), and +for+(purpose/scope). This becomes
+    # the payload of JWS. For example:
+    #
+    #   { 'pld' => 'a private message', 'for' => 'test', 'exp' => '2015-06-29T07:16:12.579Z' }
+    #
+    # JWS Signature - Signature over the JWS Header and the JWS Payload(Claims set).
+    # The signing input is given by ASCII(BASE64URL(UTF8(JWS Protected Header)) || '.' ||
+    # BASE64URL(JWS Payload)).
+    #
+    # ==== Options
+    #
+    # * <tt>:expires_at</tt> - Expiry time of the signed message.
+    # * <tt>:expires_in</tt> - Time from now after which the message
+    #   will expire (e.g. 1.month).
+    # * <tt>:for</tt> - Purpose/Scope of the signed message (defaults to 'universal').
+    #
+    # ==== Examples
+    #
+    # Messages can be signed to ensure that the data hasn't been tampered with.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
-    #   verifier.generate 'a private message' # => "BAhJIhRwcml2YXRlLW1lc3NhZ2UGOgZFVA==--e2d724331ebdee96a10fb99b089508d1c72bd772"
-    def generate(value)
-      data = encode(@serializer.dump(value))
-      "#{data}--#{generate_digest(data)}"
+    #   verifier.generate('a private message')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIOdW5pdmVyc2FsBjsGVDoIZXhwSSIdMjAxNS0wOC0wOVQxODowNjowMy44NzhaBjsGVA==.ZmRlYjhlMTkxNGI0NGQ4OWRkNDdjZWM5YTY1NmQ3MTM5OWY5NzRmMQ=="
+    #
+    # Messages can be given a purpose/scope to bump the security up some more.
+    # In this way evildoers can't reuse a sign-up form's cookie on the login
```

Comment:
`# This way evildoers can't reuse a message generated for the sign-up form on any other page:`